dodiscover/toporder/nogam.py

```diff
@@ -0,0 +1,209 @@
+        inferred topological order.
+
+        If self.do_pns = True or self.do_pns is None and number of nodes >= 20, then
+        Preliminary Neighbors Search is applied before CAM pruning.
+
+        Parameters
+        ----------
+        X : np.ndarray of shape (n_samples, n_nodes)
+            Matrix of the data.
+        A_dense : np.ndarray of shape (n_nodes, n_nodes)
+            Dense adjacency matrix to be pruned.
+
+        Returns
+        -------
+        A : np.ndarray
+            The pruned adjacency matrix output of the causal discovery algorithm.
+        """
+        d = A_dense.shape[0]
+        if (self.do_pns) or (self.do_pns is None and d > 20):
+            A_dense = self.pns(A_dense, X=X)
+        return super().prune(X, A_dense)
+
+    def _mse(self, X: NDArray, Y: NDArray) -> List[float]:
+        """Predict each column of Y from X and compute the Mean Squared Error (MSE).
+
+        Parameters
+        ----------
+        X : np.ndarray
+            Matrix of predictors observations. Usually, the n x d matrix R of
+            estimated residuals
+        Y  : np.ndarray
+            Matrix of the target variables Y[:, i]. Usually the n x d matrix D of the
+            estimated score function
+
+        Returns
+        -------
+        err : np.array of shape (n_nodes, )
+            Vector with MSE in the prediction of score_i from residual_i
+        """
+        err = []
+        _, d = Y.shape
+        err = [
+            np.mean(
+                (
+                    Y[:, col]
+                    - cross_val_predict(
+                        self._create_kernel_ridge(),
+                        X[:, col].reshape(-1, 1),
+                        Y[:, col],
+                        cv=self.n_crossval,
+                    )
+                )
+                ** 2
+            ).item()
+            for col in range(d)
+        ]
+        return err
+
+    def _estimate_residuals(self, X: NDArray) -> NDArray:
+        """Estimate the residuals by fitting a KernelRidge regression.
+
+        For each variable X_j, regress X_j on all the remaining variables  of X, and
+        estimate the residuals.
+
+        Parameters
+        ----------
+        X : np.ndarray of shape (n_samples, n_nodes)
+            Matrix of the data.
+
+        Returns
+        -------
+        R : np.ndarray of shape (n_samples, n_nodes)
+            Matrix of the residuals estimates.
+        """
+        R = []
+        R = [
+            X[:, i]
+            - cross_val_predict(
+                self._create_kernel_ridge(),
+                np.hstack([X[:, 0:i], X[:, i + 1 :]]),
+                X[:, i],
+                cv=self.n_crossval,
+            )
+            for i in range(X.shape[1])
+        ]
+        return np.vstack(R).transpose()
+
+    def _create_kernel_ridge(self):
+        return KernelRidge(kernel="rbf", gamma=self.ridge_alpha, alpha=self.ridge_alpha)
```

Comment:
Just curious. Does it make sense to preserve the KernelRidge estimator as a "fitted" property of the class?

e.g.

`self.kernel_ridge_est_ : KernelRidge`

Would a user ever want/need to query the estimator directly? Would it save efficiency? If not, it's fine. Jw.

Reply:
The regressor is created multiple times, each of which it is fitted on different data - this happens in the `self.top_order` method when calling `self._estimate_residuals` and `self._mse`.
The `self._create_kernel_ridge` is then just a shortcut that makes easy to regenerate a model with the same parameters.

Reply:
Ah I see, so the regressor is in some sense a "private attribute" and will never need to be seen by the user. Is that correct?

Reply:
Yes